### ISSUE TITLE
How exactly does ShellRegion work?

### ISSUE THREAD

**Issue by James-Ilosta** [sandialabs/pyNuMAD#19]
Hi, I'm looking through the functions at the moment and I've come across the ShellRegion function from the pyNuMAD.shell.ShellRegionClass. From what I can tell, this function defines how the face is meshed and the actual meshing occurs using the Mesh3D function. My understanding of it is that it requires a 4 sided object, defined by an array of points and the number of elements on each side is defined in the nEl parameter. What I'm wondering is:
Does this require a 4 sided object?
How does the function interpret the points provided? In all instances of its use that I've seen, the points array contains more points that are contained inside of the shape.
How does the shape handle curved edges? In the shellMeshGeneral function, it appears to mesh curved edges?
What do the points contained within the external perimeter of the shape do? Are they additional nodes for the the elements to pass through (essentially helping to structure the mesh)? Or is there another purpose for this?
What does the first argument relate to?
Is there a required order to the points?
How is the nEl parameter ordered?

Apologies for the long set of questions. I'm just trying to get to grips with the operation of the function.

ShellRegion('quad2',points,nEl,elType='quad',meshMethod='Method1')

**Comment by kbonney** [comment]
Hi @James-Ilosta. We are currently working on expanding the documentation to clarify how the meshing functionality works. Be on the lookout for those updates. In the meantime @andersone35 may be able to address some of your questions.

**Comment by andersone35** [comment]
Thanks for the question.

Yes, as Kirk said we're working toward documentation on this but it's still pending.  It was not expected that the ShellRegion class would be used directly by users in the near term.  But I can at least briefly answer these questions for now.  You're overall understanding is already remarkably good considering the lack of documentation.  

A ShellRegion represents a piece of a shell, or surface structure, like the surface of a blade.  So for the ShellRegion constructor:

ShellRegion('quad2',points,nEl,elType='quad',meshMethod='Method1')

The first input is a tag specifying the type of region.  'quad2' means a curved, 4-sided region with 2nd order interpolation defining the shape and contour through 3D space.  Currently 'quad1' (linear) and 'quad3' (cubic) are also supported.

The second input, points is a list of coordinates for the points defining the shape/placement of the region.  These point _do not necessarily become nodes of the mesh_ in the end, they just guide the interpolation of the surface within its boundaries.  For 'quad2' regions, the points should be given first with the four corners, then midpoints of sides, then center point of the quadrilateral region, like this (I hope it makes sense):

                    (side 3)
              4          7          3
              *--------*-------*
              |            |           |
(side 4)  * 8-----* 9------* 6  (side 2)
              |            |           |
              * 1-----* 5-----* 2
                     (side 1)

For 'quad1' it is just the corners of course, and 'quad3' follows a similar pattern as 'quad2'

The third input, as you guessed is the number of elements desired on sides 1 through 4 in the illustration.  This number of elements is strictly honored no matter what by the mesher, so you always have the expected number on the boundaries for continuity.

The last input is the method of mesh generation, either 'structured' or 'free'.  'structured' is good when the number of elements on opposing sides of the region are either the same or nearly the same, while free produces and unstructured mesh when createShellMesh is called.

I hope that at least clarifies some, we are always working on updates and are trying to keep up the documentation with it but it's a gradual process.  Thanks for your interest and input, we'll do our best to help out as needed.

Evan

**Comment by andersone35** [comment]
Sorry that figure came out really bad.  Here's an image:

![quad2](https://github.com/sandialabs/pyNuMAD/assets/61843712/9eb74359-30db-4e84-8e8f-10579892afdd)

**Comment by andersone35** [comment]
I also forgot, elType is just 'quad' for 4-sided elements or 'tri' for 3-sided.

**Comment by andersone35** [comment]
Yes, exactly.  To your question, in general 3D space "bottom left" really has no meaning, the important thing is just the pattern of connectivity, that the first four points go around the perimeter in one direction, and the sides, body points follow in that direction and order like the diagram shows.  It's just like defining a single shell finite element really.
